### PR TITLE
Fix locking inside TimerDescriptor

### LINKED ISSUE #37686: A query may stuck in `TimerDescriptor::drain()`
https://s3.amazonaws.com/clickhouse-test-reports/37658/486a11a5e25240c3e9b078f9a38b51e3e078acdc/stress_test__undefined__actions_.html

Stacktraces: https://pastila.nl/?008065e9/6a66e9d01a2b1479b7b0dcf8dfd5eb34

cc: @KochetovNicolai

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Due to a bug in Linux Kernel, a query can hung in `TimerDescriptor::drain`. This closes https://github.com/ClickHouse/ClickHouse/issues/37686

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit dfeb1991164bd6c8b0efc8bdcfe9dcd5b8906928 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67702/dfeb1991164bd6c8b0efc8bdcfe9dcd5b8906928/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67702/dfeb1991164bd6c8b0efc8bdcfe9dcd5b8906928/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Common/TimerDescriptor.cpp [hunk 1]
```diff
@@ -75,10 +76,22 @@ void TimerDescriptor::drain() const
     /// or since the last successful read(2), then the buffer given to read(2) returns an unsigned 8-byte integer (uint64_t)
     /// containing the number of expirations that have occurred.
     /// (The returned value is in host byte order—that is, the native byte order for integers on the host machine.)
+
+    /// Due to a bug in Linux Kernel, reading from timerfd in non-blocking mode can be still blocking.
```
Comment (yakov-olkhovskiy) [inline src/Common/TimerDescriptor.cpp:80]: do you have a link to this bug? what I have found is this:
https://patchwork.kernel.org/project/linux-fsdevel/patch/b4059ed0-5567-44e7-95f7-f7e4b227501c@kernel.dk/
but we avoiding this by breaking the loop on `EAGAIN`

**Review by yakov-olkhovskiy — APPROVED** [review]

File: src/Common/TimerDescriptor.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply src/Common/TimerDescriptor.cpp:80]: No, I didn't find it.

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/issues/67730 stood in my way.

**Comment by alexey-milovidov** [comment]
WTF: https://github.com/ClickHouse/ClickHouse/issues/54748

**Pull request merged**